File: test-suite/src/test/java/io/micronaut/views/docs/fieldset/AuthorFetcherTest.java
package io.micronaut.views.docs.fieldset;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.views.fields.Form;
import io.micronaut.views.fields.FormGenerator;
import io.micronaut.views.fields.elements.InputHiddenFormElement;
import io.micronaut.views.fields.elements.Option;
import io.micronaut.views.fields.elements.SelectFormElement;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

@MicronautTest(startApplication = false)
class AuthorFetcherTest {

    @Inject
    FormGenerator formGenerator;

    @Test
    void theOptionsOfTheSelectAreLoadedWithTheFetcher() {
        Form form = formGenerator.generate("/books/authors/save", BookAuthorSave.class);
        assertEquals("/books/authors/save", form.action());
        InputHiddenFormElement bookId = assertInstanceOf(InputHiddenFormElement.class, form.fieldset().fields().get(0));
        assertEquals("bookId", bookId.name());
        SelectFormElement authorId = assertInstanceOf(SelectFormElement.class, form.fieldset().fields().get(1));
        assertEquals("authorId", authorId.name());
        List<Option> options = authorId.options();
        assertEquals(List.of("1", "2", "3"), options.stream().map(Option::value).toList());
        assertEquals(List.of("Kishori Sharan", "Peter Späth", "Sam Newman"), options.stream().map(o -> o.label().defaultMessage()).toList());
        assertFalse(options.stream().anyMatch(Option::selected));

        form = formGenerator.generate("/books/authors/save", new BookAuthorSave(1L, 3L));
        authorId = assertInstanceOf(SelectFormElement.class, form.fieldset().fields().get(1));
        assertEquals(List.of(false, false, true), authorId.options().stream().map(Option::selected).toList());
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/views/docs/fieldset/AuthorFetcherTest.kt
package io.micronaut.views.docs.fieldset

import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import io.micronaut.views.fields.FormGenerator
import io.micronaut.views.fields.elements.InputHiddenFormElement
import io.micronaut.views.fields.elements.Option
import io.micronaut.views.fields.elements.SelectFormElement
import jakarta.inject.Inject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertInstanceOf
import org.junit.jupiter.api.Test

@MicronautTest(startApplication = false)
internal class AuthorFetcherTest {

    @Inject
    lateinit var formGenerator: FormGenerator

    @Test
    fun theOptionsOfTheSelectAreLoadedWithTheFetcher() {
        var form = formGenerator.generate("/books/authors/save", BookAuthorSave::class.java)
        assertEquals("/books/authors/save", form.action())
        val bookId = assertInstanceOf(InputHiddenFormElement::class.java, form.fieldset().fields()[0])
        assertEquals("bookId", bookId.name())
        var authorId = assertInstanceOf(SelectFormElement::class.java, form.fieldset().fields()[1])
        assertEquals("authorId", authorId.name())
        val options = authorId.options()
        assertEquals(listOf("1", "2", "3"), options.map(Option::value))
        assertEquals(listOf("Kishori Sharan", "Peter Späth", "Sam Newman"), options.map { it.label().defaultMessage() })
        assertFalse(options.any(Option::selected))

        form = formGenerator.generate("/books/authors/save", BookAuthorSave(1L, 3L))
        authorId = assertInstanceOf(SelectFormElement::class.java, form.fieldset().fields()[1])
        assertEquals(listOf(false, false, true), authorId.options().map(Option::selected))
    }
}
